[WIP] Async incremental builder

vasily-kirichenko opened the pull request.

dsyme commented on src/fsharp/vs/IncrementalBuild.fs:
```diff
@@ -504,197 +505,235 @@ module internal IncrementalBuild =
-                currentsig <> Signature(bt,id)
-            else false
+    let ForeachAction ctok (Target(output, optSlot)) bt (actionFunc:Action -> 'T -> Async<'T>) (acc:'T) : Async<'T> =
+        async {
```
I know you don't like it, but would it be ok to use 2-space indent at least until PR is finalized, to minimize the visible diff?  I know you can add the whitespace option to github, but last time I tried that it inhibited the ability to add code comments in the review.  

Thanks

vasily-kirichenko replied:
I don't understand. Everything in this file is indented with 4-spaces.

dsyme replied:
I mean changing

```
let f x = 
    let a = b
    let a = b
    let a = b
    let a = b
    c
```
to
```
let f x = 
  async {
    let a = b
    let a = b
    let a = b
    let a = b
    return c
  }
```
i.e inserting the monad using two-space and not changing most of the function (at least initially while getting the basic PR into shape).  It's just really useful to see the minimized diff and be able to comment on it directly on github

dsyme commented on src/fsharp/vs/IncrementalBuild.fs:
```diff
@@ -1773,28 +1828,34 @@ type IncrementalBuilder(ctokCtor: CompilationThreadToken, frameworkTcImportsCach
+    member builder.GetParseResultsForFile (ctok: CompilationThreadToken, filename) =
+        async {
+            let slotOfFile = builder.GetSlotOfFileName filename
+    #if FCS_RETAIN_BACKGROUND_PARSE_RESULTS
```
Align ``#if`` on column zero please

dsyme commented on src/fsharp/vs/Reactor.fs:
```diff
@@ -74,7 +74,7 @@ type Reactor() =
                         if ct.IsCancellationRequested then ccont() else
```
Not sure if this line is needed anymore

vasily-kirichenko commented:
@dsyme It seems to work. I added few `Async.RunSynchronously` to stop asyncs propagation and replaced exceptions with `Choice<'a, FSharpErrorInfo>` in a number of places. I'm not sure how much "cancellable" this version is compared to master, please review.

dsyme commented on src/fsharp/CompileOps.fs:
```diff
@@ -4291,22 +4292,29 @@ type TcImports(tcConfigP:TcConfigProvider, initialResolutions:TcAssemblyResoluti
+        return dllinfos, ccuinfos
+      }
       
     member tcImports.DoRegisterAndImportReferencedAssemblies(ctok, nms) = 
```
I suspect``DoRegisterAndImportReferencedAssemblies`` should be cancellable, so ``BuildFrameworkTcImports`` and ``BuildNonFrameworkTcImports`` are as well, but ``ImplicitLoadIfAllowed`` should not (i.e. should use RunSynchronously)

BTW one of the problems this approach may have is that the use of ``async`` may interfere with the use of thread statics for ``buildPhase`` and ``errorLogger``, as the execution of the async may reschedule continuations on other threads. 

We could perhaps use a less-rich ``cancellable { ... }`` monad that stays on the same thread, can use thread statics, but does implicit cancellation checks and goes down a cancellation continuation rather than raising an exception.  Or perhaps ``eventually { ... }`` could be adjusted to also carry a cancellation continuation.

vasily-kirichenko replied:
Done. However, this propagation stops now here https://github.com/Microsoft/visualfsharp/pull/2454/files#diff-7cfe8057cee7d7428672d69982354073R1616
I don't know what to do with it. Looks like we cannot push it forward without merging `eventually` with `async`.

dsyme replied:
Yeah I figured these would collide at some point :)

vasily-kirichenko replied:
Ah, that were not `eventually`. Got rid of `Async.RunSynchronously`.

dsyme commented on src/fsharp/vs/IncrementalBuild.fs:
```diff
@@ -375,8 +375,9 @@ module internal IncrementalBuild =
         match GetExprById(bt,seek) with 
-        | ScalarBuildRule _ ->failwith "Attempt to get width of scalar." 
-        | VectorBuildRule ve -> Option.get (GetVectorWidthByExpr(bt,ve))
+        | ScalarBuildRule _ -> 
```
remove this whitespace change, thanks

vasily-kirichenko replied:
Done.

dsyme commented on src/fsharp/vs/Reactor.fs:
```diff
@@ -83,9 +82,14 @@ type Reactor() =
+                        | Some bgOp ->
+                            let rec loop () =
+                                async {
+                                    let! r = bgOp ctok
```
An observation: the ``Step`` background action (which is what ``bgOp`` is) is never actually cancelled in the code - it's always only called as part of the Reactor agent, and the core async loop of the MailboxProcessor is never cancelled (effectively runs with CancellationToken.None).

This just makes me wonder whether either
1. it would be useful to make thee individual steps cancellable, or
2. if the ``bgOp`` should be non-Async to record the invariant that it is never cancelled.  

I'm not sure, need to think

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -2044,20 +2045,22 @@ type FSharpCheckFileResults(errors: FSharpErrorInfo[], scopeOptX: TypeCheckInfo
     member info.GetDeclarationListInfo(parseResultsOpt, line, colAtEndOfNamesAndResidue, lineStr, qualifyingNames, partialName, ?hasTextChangedSinceLastTypecheck) = 
         let hasTextChangedSinceLastTypecheck = defaultArg hasTextChangedSinceLastTypecheck (fun _ -> false)
-        reactorOp "GetDeclarations" FSharpDeclarationListInfo.Empty (fun ctok scope -> scope.GetDeclarations(ctok, parseResultsOpt, line, lineStr, colAtEndOfNamesAndResidue, qualifyingNames, partialName, hasTextChangedSinceLastTypecheck))
+        reactorOp "GetDeclarations" FSharpDeclarationListInfo.Empty (fun ctok scope -> 
```
Maybe change ``reactorOp`` to include the ``async.Return``?

vasily-kirichenko replied:
Done.

dsyme commented on src/fsharp/vs/service.fs:
```diff
@@ -2642,38 +2673,43 @@ type BackgroundCompiler(referenceResolver, projectCacheSize, keepAssemblyContent
-        let builderOpt,creationErrors,_ = getOrCreateBuilder (ctok, options, ct)
+    member private bc.ParseAndCheckProjectImpl(options, ctok) : Async<FSharpCheckProjectResults> =
+      async {
+        let builderOpt,creationErrors,_ = getOrCreateBuilder (ctok, options)
```
I'm pretty sure that ``getOrCreateBuilder`` should end up being ``Async`` (and thus cancellable) in this new world?

vasily-kirichenko replied:
Yes, it'd become async.

vasily-kirichenko commented:
@dotnet-bot test this please

majocha commented:
I'm getting
```
---- DEBUG ASSERTION FAILED ----
---- Assert Short Message ----
Unexpected exception seen in language service
---- Assert Long Message ----
System.InvalidCastException: Unable to cast object of type 'Microsoft.FSharp.Control.FSharpAsync`1[System.Object]' to type 'Microsoft.FSharp.Compiler.TypeCheckAccumulator'.
   at Microsoft.FSharp.Core.LanguagePrimitives.IntrinsicFunctions.UnboxGeneric[T](Object source)
   at Microsoft.FSharp.Compiler.IncrementalBuild.Vector.BoxingScanLeft@952TT.Invoke(CompilationThreadToken ctok, a a, b i)
   at Microsoft.FSharp.Compiler.IncrementalBuild.Action.Execute(CompilationThreadToken ctok)
   at Microsoft.FSharp.Compiler.IncrementalBuild.ExecuteApply@804-1.Invoke(Unit unitVar)
   at Microsoft.FSharp.Control.AsyncBuilderImpl.callA@841.Invoke(AsyncParams`1 args)
   at Microsoft.VisualStudio.FSharp.LanguageService.Assert.Exception(Exception e)
   at Microsoft.VisualStudio.FSharp.Editor.CommonRoslynHelpers.computation@89-2.Invoke(Exception _arg1)
   at Microsoft.FSharp.Control.AsyncBuilderImpl.callA@841.Invoke(AsyncParams`1 args)
...
...
```
when I trigger the cancellation as per the repro.

vasily-kirichenko commented:
@majocha yes, I've been trying to fix it for an hour :(

majocha commented:
It still worked great in this commit https://github.com/Microsoft/visualfsharp/pull/2454/commits/79ccf422a276dffdd72496ed29cd0162abf300a3  :(

vasily-kirichenko commented:
I know.

majocha commented:
Now after cancelling it seemingly never resumes. I don't want to make an idiot of myself again, but is it because a `IncrementalBuilder` with unresolved references is cached here 
[https://github.com/vasily-kirichenko/visualfsharp/blob/dec3f544969817913103bcb8c564093e2c88f3fc/src/fsharp/vs/service.fs#L2299](https://github.com/vasily-kirichenko/visualfsharp/blob/dec3f544969817913103bcb8c564093e2c88f3fc/src/fsharp/vs/service.fs#L2299)
and then reused each time Roslyn tries to recheck?

vasily-kirichenko commented:
@majocha you are right

vasily-kirichenko commented:
@dsyme I think I don't smart enough to do this work, sorry, Don :( You said you could do it yourself this week, do it, please.

dsyme commented:
@vasily-kirichenko I'll try to take a look tomorrow :) It's a great draft!

dsyme commented:
@vasily-kirichenko OK, I'm taking a stab in https://github.com/Microsoft/visualfsharp/pull/2460

Once it compiles green it would be great if you could give it a go

dsyme commented:
@vasily-kirichenko I'll close this in favour of the other PR.  Thanks for the draft, it was really helpful to avoid some issues (like ``ignore`` :) )

vasily-kirichenko commented:
@dsyme thank you, it's just great! :)
